### DISCUSSION TITLE
Unable to fully secure GraphQL endpoint

### DISCUSSION THREAD

**Discussion by tv42** [keystonejs/keystone#7425] [Questions] [answered]
Run a KeystoneJS Admin UI project that requires user login (`npm run dev` and `npm run build && npm run start` result in the same outcome).

Open http://localhost:3000/api/graphql (adjust host & port as needed) in browser incognito mode.
Observe that the GraphQL schema is always public.
Observe that attackers can attempt to read all information, and it's up to your hooks etc to properly prevent that. This is an unnecessary risk for APIs intended for internal consumption only.

Consider that field names and such may reveal non-public information, such as a partnership or features that haven't been announced yet.

I had hoped for:

Have a way to limit GraphQL API to authorized accounts, bearer tokens, or such.

Have visible GraphQL schema limited according to access control (not sure if this possible in GraphQL?)

**Comment by Olli1080** [comment]
If you're not overriding the [apolloConfig](https://keystonejs.com/docs/apis/config#graphql=) for introspection then it's disabled by default if NODE_ENV is set to production (see [here](https://www.apollographql.com/docs/federation/enterprise-guide/graph-security/#turn-off-introspection-in-production)).
E.g. `env NODE_ENV=production npx keystone start`
Keep in mind that this only reduces discoverability of the graphql api.

**Comment by molomby** [comment, marked as answer]
@Olli1080 is correct. The schema introspection is only enabled by default in development (specifically, when `NODE_ENV != 'production'`). It's not intended (and shouldn't be necessary) for the GraphQL schema to ever be publicly exposed.

The [GraphQL API docs](https://keystonejs.com/docs/apis/graphql#using-the-api) describes how to take control of this config if you want to , say, expose it in production.

> Note that for security, both the playground and [introspection](https://graphql.org/learn/introspection/) are disabled when running Keystone in production. You can control this behaviour using the `config.graphql.playground` and `config.graphql.apolloConfig` options. For example, to disable these features in dev environments, add this to your Keystone config: `graphql: { playground: false, apolloConfig: { introspection: false } }`.

As @Olli1080 also points out, hiding these features (the playground and introspection) _helps_ with security but is arguably just obfuscation – it certainly shouldn't be relied on to protect your system. For that, you do need solid [access control](https://keystonejs.com/docs/guides/auth-and-access-control#adding-access-control) either configured for your lists/fields or built into your hooks, custom mutations, etc.

I should maybe also mention that you can configure fields to be omitted from the GraphQL API entirely using `graphql: 'omit'` (see the [common field config](https://keystonejs.com/docs/apis/fields#common-configuration)). This is a pretty extreme measure – fields hidden in this way will be completely missing from GraphQL and, of course, the Admin UI – but remain available though the [DB API](https://keystonejs.com/docs/apis/db-items) and will still be managed by Keystone in terms of migrations. This can be handy for internal or security sensitive fields that are only accessed in custom code within your app (hooks, custom mutations, etc).

Oh, one other thing – if you're especially concerned with leaking meta data, it should be possible to enable [persisted queries](https://www.apollographql.com/docs/apollo-server/performance/apq/) on the Apollo instance Keystone uses. This further obscures your schema by replacing the GraphQL queries you use in production with hashes, further obscuring your data schema.
